Allow select all/ deselect all when hiding dataframe columns from UI

plumol opened the pull request:
## Describe your changes
Introduces a 'Select all/ deselect all' checkbox when hiding/ showing dataframe columns

<!-- If it's a visual change, please include a screenshot or video! -->

https://github.com/user-attachments/assets/c22fbd79-4c2a-41d9-bb6b-ad616ed14204

### Preview of the select all feature
<img width="747" height="277" alt="select all preview" src="https://github.com/user-attachments/assets/668b49a0-909c-4adc-88e9-7edc6a1a7821" />

When deselected, all other columns are deselected and are hidden from view.
<img width="747" height="277" alt="deselect all preview" src="https://github.com/user-attachments/assets/009fb1ff-7519-4b0c-bf36-bc7d8a9976be" />

## GitHub Issue Link (if applicable)

- Addresses the main functionality of #12082

## Testing Plan

- Unit Tests (JS and/or Python)
- - Added some preliminary unit tests already, but I appreciate any feedback or additional extensions of them
- E2E Tests
- Any manual testing needed?
- - UI formatting on looks? Should this be separated by a border from the rest of the column options?

I am a first time contributor, so I'd really appreciate any feedback on style and other areas I can improve on.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

plumol committed:
Allow select all/ deselect all when hiding dataframe columns from UI

plumol committed:
Update frontend/lib/src/components/widgets/DataFrame/menus/ColumnVisibilityMenu.test.tsx

Co-authored-by: graphite-app[bot] <96075541+graphite-app[bot]@users.noreply.github.com>

sfc-gh-bnisco commented:
Thank you for your contribution @plumol! Tagging our dataframe expert @lukasmasuch to take a look at this.

lukasmasuch commented:
Thanks for the contribution 👍  I think it is a good idea to add a "select all" functionality. But maybe we need some kind of divider to separate it from the actual columns. It could be as simple as just adding a divider similar to here: 

<img width="266" height="235" alt="Screenshot 2025-08-14 at 01 57 47" src="https://github.com/user-attachments/assets/6039bb46-f351-480f-ac4c-e11bc2c1955d" />

It also might be nice to add a search box eventually. But doesn't need to be in the same PR.

plumol commented:
Thanks for the review and comments! I'll work on addressing the issues brought up by the bots and add a divider for the select all box. 

Something that I noticed is that the visibility menu is not enabled by default and is automatically hidden when there are no hidden columns. This is makes the select all behavior clunky to deal with as you have to manually re-hide a column first to enable the visibility button (you can see an example in the attached video). Do you think it's a good idea to enable visibility controls by default or leave as is?

plumol committed:
Added select all divider, refactored select logic to utilize baseui checkbox properties better

plumol commented:
Renders and video of features. Currently, select-all adopts 3 different states: checked, indeterminate, and unchecked. The indeterminate state utilizes baseui/checkbox 's indeterminate property. 

Clicking:
1. A checked state will deselect all items, 
2. An indeterminate state will deselect all items 
3. An unchecked state will select all items.

The select all box is stickied to the top of the visibility menu. I feel like the dark mode version is... a bit ugly no? I think the contrast is a bit too much but open to feedback on the other details too.

https://github.com/user-attachments/assets/fcb252bf-1b81-4618-9c84-9f78c398eea4

<img width="747" height="405" alt="light mode render" src="https://github.com/user-attachments/assets/e7be55a9-539e-42a3-8186-9128262a5233" />

<img width="786" height="454" alt="dark mode render" src="https://github.com/user-attachments/assets/197d932f-20eb-4b76-869a-ae8752d4a1aa" />

plumol committed:
Reverted back to original select all functionality, added sticky header for visibility menu

plumol committed:
Cleaned commented code out

plumol committed:
Merge branch 'streamlit:develop' into feature/dataframe-column-select-all-visibility

plumol committed:
Updated playwright screenshots for new select all

plumol commented:
Hi @lukasmasuch! Just wanted to bump this after the weekend. Thanks!

lukasmasuch committed:
Merge remote-tracking branch 'upstream/develop' into pr/plumol/12164

lukasmasuch commented:
> Do you think it's a good idea to enable visibility controls by default or leave as is?

We did that to keep the dataframe toolbar a bit cleaner. But not something set in stone, but I think it's better to keep it as is in this PR. I will discuss it internally if we want to always show the action.

lukasmasuch commented:
> An indeterminate state will deselect all items

Hmm, I think it might be better to select all items in this case since unselecting all isn't that useful (compared to selecting all). Is that possible to change?

> The select all box is stickied to the top of the visibility menu. I feel like the dark mode version is... a bit ugly no? I think the contrast is a bit too much but open to feedback on the other details too.

Can we use the same background color for the top (`theme.colors.secondaryBg`)?

plumol committed:
Revert select all functionality to select on indeterminate and changed background color of list

plumol commented:
> Hmm, I think it might be better to select all items in this case since unselecting all isn't that useful (compared to selecting all). Is that possible to change?
> 
> >

Sure, that's a quick fix to the selection logic above. I think in either case, you lose some functionality because you can't access the visibility menu by default. 

If the indeterminate state:
1) Deselects all, you have to click twice to have all items selected (once to deselect all items to the unchecked state, and another time to reselect all items)
2) Selects all, you only have to click once to have all items selected, but you lose out on the 'deselect all' feature entirely since the visibility menu is hidden when all items are visible and selected.

The other gripe with it is the travel and user mouse movement when re-enabling the visibility menu- you have to click onto a new column and travel down the menu to re-click 'hide column'. Not really a groundbreaking issue, but I can see how it can be annoying to re-open the menu if you accidentally hide it by 'selecting all'.

plumol committed:
Removed extra conditional check and removed background color override on Select All box to remain consistent with light and dark themes

plumol committed:
Fixed playwright e2e test for text, passes local e2e tests as well

lukasmasuch committed:
Merge branch 'feature/dataframe-column-select-all-visibility' of https://github.com/plumol/streamlit into pr/plumol/12164

lukasmasuch committed:
Apply some tweaks

lukasmasuch commented:
> but you lose out on the 'deselect all' feature entirely since the visibility menu is hidden when all items are visible and selected.

I think that's fine since the deselect all state is a bit confusing (just showing the index).

> The other gripe with it is the travel and user mouse movement when re-enabling the visibility menu- you have to click onto a new column and travel down the menu to re-click 'hide column'. Not really a groundbreaking issue, but I can see how it can be annoying to re-open the menu if you accidentally hide it by 'selecting all'.

Yeah, I think we can "fix" this by always showing the visibility menu toolbar action. I will bring this up internally.

lukasmasuch reviewed (approved):
Thanks for the contribution 👍  I added a couple of small tweaks to merge it today in time for the 1.49 cut-off.

lukasmasuch committed:
Update snapshots

plumol commented:
Thanks for the help! I'll see what else I can contribute to, maybe starting with the search box feature mentioned earlier in the thread. 😃

aimtsou commented:
@plumol, @lukasmasuch 

By the way, not allowing deselect all is not a good idea. Suppose that I have a dataframe with 200 columns but I want to select only 2 or 3 and get the csv. Without allowing the deselection of columns now this is not possible, I have to make 200 clicks. When deselecting, allowing the index only is better.

plumol commented:
@aimtsou That's one of the concerns that I had earlier when I was figuring out how what behavior to default the 'select all' to. The deselect feature does exist, but it's not technically accessible because the column visibility menu is automatically hidden when all columns are visible. 

The fix to this with the current behavior would be to always enable the column visibility menu, but because of some design choices (for which I don't know the full context of because I'm just an open-source contributor) this isn't the case. I brought up some of these points in the earlier conversations if you'd like to read them. In any case, I do agree that this is something that might need more discussion.
